### PR TITLE
Unify internal headers in library/

### PR DESCRIPTION
**Pull request by chris-jones-arm**
## Description
Move all internal headers, currently in `include/mbedtls/*_internal.h` and `library/*.h`, to a new folder: `include/internal/`. This should make it easier to maintain and organise internal headers for the future and allow them to be edited without a new major release as these headers are internal.

Implements #4032.

## Status
**READY**

## Requires Backporting

NO  

This is part of mbed TLS 3.0 and includes API breaking changes and is not being backported.

## Todo

- [x] Move all files to `library/`
- [x] Rename all to `*_internal.h`
- [x] Add bn_mul.h to `library/`
- [x] Change documentation for changed filename/label as internal

### CODE REVIEW

File: 3rdparty/everest/library/Hacl_Curve25519_joined.c [hunk 1]
```diff
@@ -19,7 +19,7 @@
  *  This file is part of mbed TLS (https://tls.mbed.org)
  */
 
-#include "common.h"
+#include "internal/common.h"
```
Comment (gilles-peskine-arm) [inline 3rdparty/everest/library/Hacl_Curve25519_joined.c:22]: As I wrote in https://github.com/ARMmbed/mbedtls/issues/4032#issuecomment-782201943, I don't think we should move headers from `library` to `include/internal`. I see two practical reasons:

* It would make sense to have a directory called `include/mbedtls/internal` containing header files which have to be published for the sake of the C compiler, but that application code should not depend on. (For example: struct definitions where we want the compiler to know the size, but we don't want application code to access fields directly.) Having `include/internal` with a different meaning is confusing.
* Installers can currently copy `include/*` to the system include path. `include/internal` breaks this.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
There was no agreement to creating `include/internal` in https://github.com/ARMmbed/mbedtls/issues/4032 and I do not think it is a good idea.

File: 3rdparty/everest/library/Hacl_Curve25519_joined.c (hunk 1 above)
Comment (chris-jones-arm) [reply 3rdparty/everest/library/Hacl_Curve25519_joined.c:22]: Right sorry, I misunderstood and thought you were only referring to moving source files in that comment.

I agree with your above statement that `include/mbedtls/internal` would be a more suitable place for the internal headers, especially as it would presumably also allow us to split the PSA headers into `include/psa/internal`.

In your comment above it sounds like you would prefer to keep some header files out of `include/<mbedtls/psa>/internal` altogether. Can you clarify whether that is the case and if so where should headers that don't belong be put? For example: `ssl_invasive.h` looks like it might not fit into `include/mbedtls/internal` however I also don't agree with it being left in `library`, which at least to me seems like it should only contain source files (although I often come from a background of having a `src/` folder).

**Comment by mpg** [comment]
I agree with Gilles, I don't think `include/internal` is a good place for those headers, and I think `library` is a fine place for them.

We've been putting new internal headers directly in `library` recently, but just haven't moved pre-existing headers to the same place. I understood #4032 to be about doing just that (align old internal headers to the current scheme), not changing the scheme we're currently using for new internal headers. (Though to be fair the description didn't quite reflect that.)

File: 3rdparty/everest/library/Hacl_Curve25519_joined.c (hunk 1 above)
Comment (chris-jones-arm) [reply 3rdparty/everest/library/Hacl_Curve25519_joined.c:22]: Right, thanks for the clarification there.

Upon re-thinking I'm not against having headers inside `library/`, my main concern is unifying them in one location or at least having them in some sort of logical order. When I first cloned the library my first thought was that all the headers will be in `include` and the source files in `library/`. To me it seems strange and confusing that there are internal headers **and** public headers in `include/` **as well as** internal headers inside `library/`. This is made doubly so when the headers inside `library/` are are not named `*_internal.h` or have any mention of being only internal in the file itself (e.g: `ssl_tls13_keys.h`).

I don't think I have a particularly strong opinion on exactly where in `library/` they all go (I slightly prefer `library/<internal/include>` and then `library/*_internal.h`) but at the very least it should be documented and obvious that the files are part of the internal library and should not be used by application code.

Comment (chris-jones-arm) [reply 3rdparty/everest/library/Hacl_Curve25519_joined.c:22]: Thanks for the input @mpg, I was not aware there was a change in how internal headers are done which explains the discrepancy. I feel as though this makes the case that this is something that should be better documented so there is less reliance on knowing the history of the library.

Comment (gilles-peskine-arm) [reply 3rdparty/everest/library/Hacl_Curve25519_joined.c:22]: > When I first cloned the library my first thought was that all the headers will be in include and the source files in library/.

I'm puzzled as to why you thought this. It isn't a particularly common convention to have all header files, both public and private, under a single toplevel directory. Looking at a few examples:

* BearSSL: public headers in `inc`, private headers in `src`
* OpenSSL: public headers in `include`, private headers under `crypto` and `ssl` (and others).
* GnuTLS: public headers in `lib/includes`, private headers in `lib` (outside `lib/includes`) and `src` and others
* WolfSSL: no toplevel directory for include files, at first glance it looks like all headers in various library subdirectories are public and there are no private headers.
* Libsodium: public headers in `src/libsodium/include`, private headers in various other subdirectories of `lib/libsodium`.
* Curl: public headers in `include`, private headers in `lib` and `src`.

All of these include header files in the same directory as `.c` files. So I dispute the assertion that there's anything surprising about having header files in `library`.

> To me it seems strange and confusing that there are internal headers and public headers in include/ as well as internal headers inside library/.

But that's precisely what we wanted to fix! By moving the internal headers out of `include`. And into `library`.

**Comment by mpg** [comment]
Note: even though it doesn't have `_internal` in its name, `bn_mul.h` is also internal and should also be moved to `library`. (It's only defining macros with inline assembly, hopefully nobody ever considered that part of the public API.)

File: tests/scripts/list-identifiers.sh [hunk 2]
```diff
@@ -47,9 +47,9 @@ done
 
 if [ $INTERNAL ]
 then
-    HEADERS=$( ls include/mbedtls/*_internal.h library/*.h | egrep -v 'compat-1\.3\.h|bn_mul' )
+    HEADERS=$( ls library/*.h )
```
Comment (mpg) [inline tests/scripts/list-identifiers.sh:50]: Note: previously `bn_mul` was excluded (here and in the else branch). I'm not sure why that was (should have left a comment), so I'm not sure if it's OK to stop excluding it... Was that change intentional?

Note 2: I'm not mentioning `compat-1.3` as it's being removed by #4150

File: library/psa_crypto_its_internal.h [hunk 3]
```diff
@@ -0,0 +1,149 @@
+/** \file psa_crypto_its_internal.h
```
Comment (mpg) [inline library/psa_crypto_its_internal.h:1]: Note: github shows `psa_crypto_its_internal.h` being created, but not `psa_crypto_internal.h` being removed (or renamed to the other). That doesn't match the commit message.

File: programs/test/cpp_dummy_build.cpp [hunk 4]
```diff
@@ -100,6 +93,13 @@
 #include "mbedtls/x509_crt.h"
 #include "mbedtls/x509_csr.h"
 #include "mbedtls/xtea.h"
+#include "../library/bn_mul_internal.h"
```
Comment (mpg) [inline programs/test/cpp_dummy_build.cpp:96]: I don't think those should be here. The point of this program is to check that you can build a C++ program linking to the library and `#include`ing its public headers. Those headers are no longer public, so there's no reason to test that they can be `#include`d from C++ (we don't support building the library itself as C++).

**Review by mpg — CHANGES_REQUESTED** [review]
I left a few comments on specific points, but I also have a more general concern about the naming plan.

After moving `include/mbedtls/*_internal.h` to `library`, there's a naming discrepancy: internal headers that were originally in `include` and later moved to `library` have `_internal` in their names while those that we directly created in `library` don't. I fully agree this discrepancy is undesirable and that either all or none of the internal headers should have `_internal` in the name, but there are obviously two ways of resolving this discrepancy and I was expecting the other one: align the naming of the "old" headers to that of the "new" ones.

The commit message says it "makes it clearer when including these files that they are internal and should not be used by applications." but I think that's already clear enough from the location of the files. (Obviously when internal headers were in the same directory as public ones, `_internal` was useful to distinguish.) We've had internal headers in `library` without `_internal` in there name for a while now and I've not heard of any confusion resulting from it - do you have any concrete reason to think the current naming scheme for internal headers was not appropriate?

File: tests/scripts/list-identifiers.sh (hunk 2 above)
Comment (mpg) [reply tests/scripts/list-identifiers.sh:50]: I double-checked, and I think removing `bn_mul` here is OK.

**Comment by chris-jones-arm** [comment]
After I made the changes I realised that I too didn't like having the `_internal` suffix on internal headers. The only reason why I haven't already removed it is because some internal headers have naming collisions with other public headers (e.g: `rsa_internal.h` and `rsa.h`).

From a technical perspective that means that the header guards will conflict which can be fixed by leaving the internal one as `MBEDTLS_xxx_INTERNAL_H` but that does result in a discrepancy with the filename.

Secondly it could be confusing to have two headers with conflicting names when including them. In a file if you do `#include "rsa.h"` would you expect the internal header or the public header to be included? As it stands this is possible to figure out but it is slightly ambiguous. This ambiguity could be reduced by using a internal/include folder in `library/` so that you can have:
```C
#include "internal/rsa.h"
#include "mbedtls/rsa.h"
```
but this may just increase the complexity unnecessarily.

 I'm not against any of the above options (including just removing the suffix and leaving as is) however it's something to think about and I'm interested in your opinion on this.

File: 3rdparty/everest/library/Hacl_Curve25519_joined.c (hunk 1 above)
Comment (mpg) [reply 3rdparty/everest/library/Hacl_Curve25519_joined.c]: Thanks Gilles for making this list. Regarding the other point of discussion (whether putting a header outside of `include` is enough to identify it as internal) I checked BearSSL, OpenSSL, GnuTLS, libsodium, and none of them felt the need to add `_internal` or anything else to the names of headers outside of the public include directory in order to mark them as internal: the fact that they're out of the pubic include directory is enough. (Generally speaking, I'm sure I've see this pattern in other C projects, so to me at least that seems like a widespread and unsurprising convention.)

Coming back to an earlier point in the discussion, there are in principle 3 kinds of headers:

1. Fully public headers: application code can use what they declare.
2. Semi public headers: the compiler needs to see this when building the application, but application developers should not use that's declared there.
3. Private headers: the compiler only needs to see this when building the library, and only the library should use what's declared there.

As Gilles said, only fully and semi public headers have any business being in `include` (as long as `make install` copies all of it). Currently we only have fully public and private headers. As Gilles said, if we ever have semi-public headers, I think their natural place is in `include/mbedtls/internal` or any other place under `include` with `internal` in the path. (Btw, OpenSSL does something similar here: it has `include/openssl` for fully public headers and `include/internal` for semi-public ones.)

Now regarding private headers, before this PR we had two groups:

- 3a. those in `include/mbedtls`, most often with `_internal` in the name but not always (see `bn_mul.h`), whose location+name would perhaps be appropriate for (semi-)public headers
- 3b. those in `library`, with nothing special in the name.

The goal of this PR, that's a point we all agree on, is to only have one group of private headers.

It was and still is my opinion that the most simple and best way to achieve that is to move 3a to 3b. Instead you felt the need to move both 3a and 3b to a new 3c "library/*_internal.h". I don't see a need for that, I think 3b is the most common convention and that it's a good one.

File: tests/scripts/list-identifiers.sh (hunk 2 above)
Comment (chris-jones-arm) [reply tests/scripts/list-identifiers.sh:50]: I thought it was being specifically included not excluded so no in that sense I thought the new statement would catch the same files as before (as bn_mul.h) is now internal in `library/`. However looking at it I'm not sure why it was ever excluded in the first place (git blames you: d1ddd29e81) as it's just defining a lot of macros. Is there a reason why it needs to be excluded from the list of identifiers?

File: programs/test/cpp_dummy_build.cpp (hunk 4 above)
Comment (chris-jones-arm) [reply programs/test/cpp_dummy_build.cpp:96]: Ah okay I wasn't aware of the purpose of the file so I just made the change to make it build properly, I will remove it now.

**Comment by mpg** [comment]
Good point about name conflicts. In an ideal world, this shouldn't be a problem for `#include` as `rsa.h` and `mbedtls/rsa.h` are different - but OTOH some projects using Mbed TLS have peculiar requirements such as file names being unique across the build tree (cough Mbed OS cough). As you say, there's also the question of the header guard. So I think it's probably safer to try maintaining unique names. (Note that placing headers in a specific sub-directory of `library` wouldn't solve any of those problems.)

I'm tempted to say that the best way to achieve that is by having more precise names. For example, (writing while thinking) the internal RSA header contains helpers for RSA (mostly for importing and checking keys) that are independent of the context, so that suggests a few possible names: `rsa_helpers.h`, `rsa_import_helpers.h`, `rsa_context_indep.h`... I think those would probably be more helpful than `_internal`.

Regarding PK, MD and Cipher, the headers were previously called `pk_wrap` and so on, because they declare wrapper functions for the underlying module. (I renamed them to `_internal` in an attempt to make it clear to users that they were not part of the public API, as we didn't support header in `library` in our build system at the time.)

Regarding `ssl_internal.h` it becomes harder, as it's currently a mixed bag of things whose only common point is indeed that they don't need to be seen by users... Ideally over time its contents would be dispatched in other files (eg `ssl_msg.h` etc.) but in the meantime I'm not sure.

Regarding `ecp_internal.h` it's actually meant to be used by writers of alternative implementations of parts of the ECP module. Here I'm not even sure we should move it to `libary` after all.

Wdyt?

File: tests/scripts/list-identifiers.sh (hunk 2 above)
Comment (mpg) [reply tests/scripts/list-identifiers.sh:50]: No, in the meantime I double-check and found no reason to exclude it: excluding it may make it easier to pass `check-names.sh`, but since it now passes without excluding it (as the CI results for this PR show), I don't think there's any reason to exclude it any more.

(Said otherwise, after looking into it a bit more I withdraw this point, but I didn't want to remove the comment in case you saw it in the meantime and got confused by disappearing comments. But I've marked the conversation as resolved.)

**Comment by mpg** [comment]
Question especially for @hanno (who wrote `rsa_internal.h`) and @yanesca (who wrote `ecp_internal.h`): given the special status of these headers (meant to be used by developers of alt implementations), do you think they should go in `libary` or remain in `include`?

It seems to me that developers of alt implementation have a special position that is not the same as that of a normal user of the library, so I think it would be OK to keep then in `library`, probably with explicit names such as `xxx_alt_helpers.h`, but perhaps I'm missing something? Or perhaps it would be better to group them in some specific place (`include_alt`)? Wdyt?

**Comment by chris-jones-arm** [comment]
I agree with the best way to solve this is to have all unique names that are more precise. For example I like the idea of `rsa_helpers.h` and also `<pk/md/cipher>_wrap.h`.

For `ssl_internal.h` I don't know because I really don't know what the file contains/does. Perhaps `ssl_helpers.h` would work for now unless another name is more appropriate?

After writing 3 different answers to the second question, I've realised that I don't really know where files like `rsa_helpers.h` belong. The problem is that to alt devs they are public files but to general purpose devs they are internal files. To me this means that they don't really belong in `library/` or in `include/` but I don't think `include_alt/` is appropriate because it sounds like it is *only* available to alt implementations (it is also being used by the library). Maybe `include_helpers/` or `include/mbedtls/helpers/` would be better?

**Comment by yanesca** [comment]
`ecp_internal.h` should go into library, it was in include in the first place because we didn't have internal headers in the library back then.

**Comment by yanesca** [comment]
> so I think it would be OK to keep then in library, probably with explicit names such as xxx_alt_helpers.h, but perhaps I'm missing something? Or perhaps it would be better to group them in some specific place (include_alt)? Wdyt?

I am in favour of giving them more explicit names. I don't know if we have decided yet about adding a directory structure to `library`, so far we used prefixes in file names to organise and I wouldn't want to change that just for the alt headers.

**Comment by mpg** [comment]
Thanks @yanesca for confirming that `library` is an appropriate place for headers used by alt implementations. That was my intuition (alt implementations sort of modify the library, so they're closer to the library than to the application), but it's good to hear from someone who's more familiar than me with alt implementations.

However, if we allow alt implementers to look inside `library`, we probably want to make it clear than not every header in `library` is fair game: some are just for us. So I'd suggest using `alt` in the names of those for alt implementer (and only in those names), for example `rsa_alt_helpers.h`. Anyway, this whole "alt implementation" thing is on its way out, so we probably don't need to think about it too hard, it just needs to remain usable until it's removed.

We seem to be converging on unique filenames, and achieving said uniqueness by using descriptive names.

Regarding the case of `ssl_internal.h`, I think the honest name for it now would be `ssl_misc.h` and hopefully in the future we can dispatch its contents to other files with better names.

**Comment by chris-jones-arm** [comment]
@mpg You mention that alt is on it's way out, did you want to document that on https://github.com/ARMmbed/mbedtls-docs/pull/5/files or review, merge it and add to it afterwards? As it's the first I've heard of it, it would be good to write this down for others to find.

**Comment by mpg** [comment]
Sorry for coming up with new work items, but I think the following headers should be moved to `library` as well: `aesni.h` and `padlock.h`. Both of them have a warning in the file-level documentation saying they're internal, but didn't have `_internal` in their name.

I'm unsure about `entropy_poll.h`: it has no warning, but it looks to me like these are the raw entropy sources that should never be used directly by the user, but are only meant to be used by the entropy mixer (`entropy.h`). @gilles-peskine-arm I think at this point you're more familiar than me with the topic, wdyt?

Note: I went over all existing headers and those are the only ones I could find.

**Comment by gilles-peskine-arm** [comment]
`entropy_poll.h` is closer to `platform.h` and `ecp_internal.h`: it's meant to be used if you customize the library for your platform, not by application code as such. Since `ecp_internal.h` is moving, I guess `entropy_poll.h` should move too.

**Comment by gilles-peskine-arm** [comment]
I'll add ALT removal to https://github.com/ARMmbed/mbedtls-docs/pull/5.

File: 3rdparty/everest/library/Hacl_Curve25519_joined.c (hunk 1 above)
Comment (chris-jones-arm) [reply 3rdparty/everest/library/Hacl_Curve25519_joined.c]: Internal headers have been unified under `library/`. Would you approve this design now?

File: library/psa_crypto_its_internal.h (hunk 3 above)
Comment (chris-jones-arm) [reply library/psa_crypto_its_internal.h]: Fixed.

File: programs/test/cpp_dummy_build.cpp (hunk 4 above)
Comment (chris-jones-arm) [reply programs/test/cpp_dummy_build.cpp]: Done.

**Comment by chris-jones-arm** [comment]
PR has been reworked to address comments and also includes the new work from https://github.com/ARMmbed/mbedtls/pull/4164#issuecomment-792631308. It has also been rebased to fix the conflicts.

File: include/mbedtls/md.h [hunk 5]
```diff
@@ -79,7 +79,7 @@ typedef enum {
 #endif
 
 /**
- * Opaque struct defined in md_internal.h.
+ * Opaque struct defined in md_wrap.h.
```
Comment (gilles-peskine-arm) [inline include/mbedtls/md.h:82]: This is public documentation, but there's no `md_wrap.h` referenced in the public documentation. At least
```suggestion
 * Opaque struct defined in `library/md_wrap.h`.
```

Better, change the text in the Doxygen comment to just say that it's an opaque struct and preferably (but getting out of scope) reference the functions that apply to it (`mbedtls_md_info_from_xxx` to construct one, `mbedtls_md_get_xxx` to access the contents), and add a non-Doxygen comment referencing `library/md_wrap.h`.

This goes for the other places where headers that are now in `library` are still referenced in a comment (Doxygen or not) under `include`.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
I'm happy with the set of moved headers and where they've been moved.

There are just a few minor problems, mostly related to dangling references in documentation.

File: include/psa/crypto_extra.h [hunk 6]
```diff
@@ -297,7 +297,7 @@ void mbedtls_psa_get_stats( mbedtls_psa_stats_t *stats );
  * \param[in] seed          Buffer containing the seed value to inject.
  * \param[in] seed_size     Size of the \p seed buffer.
  *                          The size of the seed in bytes must be greater
- *                          or equal to both #MBEDTLS_ENTROPY_MIN_PLATFORM
+ *                          or equal to both MBEDTLS_ENTROPY_MIN_PLATFORM
  *                          and #MBEDTLS_ENTROPY_BLOCK_SIZE.
```
[the remark this answers is not in the document]
Comment (chris-jones-arm) [reply include/psa/crypto_extra.h]: Done.

File: programs/ssl/ssl_context_info.c [hunk 7]
```diff
@@ -48,7 +48,7 @@ int main( void )
 #include "mbedtls/error.h"
 #include "mbedtls/base64.h"
 #include "mbedtls/md.h"
-#include "mbedtls/md_internal.h"
+#include "../../library/md_internal.h"
```
[the remark this answers is not in the document]
Comment (chris-jones-arm) [reply programs/ssl/ssl_context_info.c]: Done.

File: programs/test/selftest.c [hunk 8]
```diff
@@ -56,6 +55,7 @@
 #include "mbedtls/ecjpake.h"
 #include "mbedtls/timing.h"
 #include "mbedtls/nist_kw.h"
+#include "../library/entropy_poll.h"
```
[the remark this answers is not in the document]
Comment (chris-jones-arm) [reply programs/test/selftest.c]: Done.

File: include/mbedtls/config.h [hunk 9]
```diff
@@ -2400,7 +2400,7 @@
  *          library/ecp.c
  *          library/ecdsa.c
  *          library/rsa.c
- *          library/rsa_internal.c
+ *          library/rsa_alt_helpers.h
```
[the remark this answers is not in the document]
Comment (chris-jones-arm) [reply include/mbedtls/config.h]: Done.

File: include/mbedtls/md.h (hunk 5 above)
Comment (chris-jones-arm) [reply include/mbedtls/md.h]: Done.

**Review by mpg — APPROVED** [review]
Looks good to me!

**Comment by mpg** [comment]
I'm not sure we considered the question: should this have a ChangeLog entry? On one hand, hopefully nobody was using those internal headers, on the other hand, people tend to be creative, and a complete list of changes never hurts. So I'm slightly inclined to say yes.

(Labeling "needs: changelog" provisionally until we reach an agreement one way or the other.)

**Comment by chris-jones-arm** [comment]
I had the same thoughts as you @mpg. As we have also moved headers which weren't directly labelled as internal like bn_mul.h I will make a changelog just to be on the safe side.

**Review by mpg — APPROVED** [review]
The ChangeLog entry looks good to me, thanks!

**Comment by gilles-peskine-arm** [comment]
This doesn't affect well-behaved applications, but it does affect alt implementations, so it should have a changelog entry targeted at them.

File: ChangeLog.d/move_internal_headers.txt [hunk 10]
```diff
@@ -0,0 +1,6 @@
+API changes
+   * Move internal headers.
+     All internal headers have been moved to library/ to unify them in one
+     location that is separate from the public API. This includes some files
+     that were previously not explicitly internal such as: `bn_mul.h`,
+     `aesni.h`, `padlock.h` and `entropy_poll.h`.
```
Comment (gilles-peskine-arm) [inline ChangeLog.d/move_internal_headers.txt:6]: “Move internal headers” is meaningless on its own. Remember that the audience of the changelog is user of the library. They don't know or care about our long-term plans for the library, our motivation for the change, or the context in which the change is made.

Suggested text:

> Header files that were only meant for the library's internal use and were not meant to be used in application code have been moved out of the include/ directory. The headers concerned are bn_mul.h, aesni.h, padlock.h, entropy_poll.h and *_internal.h.

(Also: this is plain text, not markdown.)

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
LGTM apart from the changelog entry.

File: ChangeLog.d/move_internal_headers.txt (hunk 10 above)
Comment (chris-jones-arm) [reply ChangeLog.d/move_internal_headers.txt]: Done along with another changelog entry for alt implementors.

**Review by gilles-peskine-arm — APPROVED** [review]

**Review by mpg — APPROVED** [review]
Thanks for improving the ChangeLog entries. Looks all good to me!

**Comment by chris-jones-arm** [comment]
Thanks for the catch, annoyingly I even wrote myself a note to remove that reference but I seem to have forgotten about it somewhere along the line. #4281 now removes that reference.
